SCClassLibrary/Common/Control/GraphBuilder.sc

```diff
@@ -56,28 +56,41 @@ NamedControl {
 	var <name, <values, <lags, <rate, <fixedLag;
 	var <control;
 
-	*ar { arg  name, values, lags;
-		^this.new(name, values, \audio, lags, false)
+	*ar { arg  name, values, lags, spec;
+		^this.new(name, values, \audio, lags, false, spec)
 	}
 
-	*kr { arg  name, values, lags, fixedLag = false;
-		^this.new(name, values, \control, lags, fixedLag)
+	*kr { arg  name, values, lags, fixedLag = false, spec;
+		^this.new(name, values, \control, lags, fixedLag, spec)
 	}
 
-	*ir { arg  name, values, lags;
-		^this.new(name, values, \scalar, lags, false)
+	*ir { arg  name, values, lags, spec;
+		^this.new(name, values, \scalar, lags, false, spec)
 	}
 
-	*tr { arg  name, values, lags;
-		^this.new(name, values, \trigger, lags, false)
+	*tr { arg  name, values, lags, spec;
+		^this.new(name, values, \trigger, lags, false, spec)
 	}
 
-	*new { arg name, values, rate, lags, fixedLag = false;
+	*new { arg name, values, rate, lags, fixedLag = false, spec;
 		var res;
 
+		this.initDict;
+
 		name = name.asSymbol;
 
-		this.initDict;
+		if (spec.notNil) {
```

Comment:
One more thing I noticed: When this method fails with an error, the spec will have been set.

Reply:
My suggestion would be the following:

Just keep this:
```supercollider
if (spec.notNil) {
			spec = spec.asSpec;

			if (values.isNil) {
				values = spec.default;
			};

};
```

But add these methods to `NamedControl`:

spec {
			^UGen.buildSynthDef.specs[name]
}

spec_ { |spec|
			spec = spec.asSpec;
			this.spec !? _.setFrom(spec) ?? {
				UGen.buildSynthDef.specs[name] = spec;
			};
}
```

and then just set the spec here https://github.com/supercollider/supercollider/blob/26038936c50968c33c223c7e6852c3a5634f38ac/SCClassLibrary/Common/Control/GraphBuilder.sc#L120

`this.spec = spec`

Reply:
Good suggestions, just fixed.